ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Monitor.hs

```diff
@@ -70,27 +72,44 @@ targetPeers :: (MonadSTM m, Ord peeraddr)
             => PeerSelectionActions peeraddr peerconn m
             -> PeerSelectionState peeraddr peerconn
             -> Guarded (STM m) (TimedDecision m peeraddr peerconn)
-targetPeers PeerSelectionActions{readPeerSelectionTargets}
+targetPeers PeerSelectionActions{ readPeerSelectionTargets = readTargets,
+                                  currentTargets,
+                                  churnMutex }
             st@PeerSelectionState{
               publicRootPeers,
               localRootPeers,
               targets,
-              ledgerStateJudgement,
+              ledgerStateJudgement = lsj,
               bootstrapPeersFlag,
-              hasOnlyBootstrapPeers
+              hasOnlyBootstrapPeers,
+              useGenesisFlag
             } =
     Guarded Nothing $ do
-      targets' <- readPeerSelectionTargets
-      check ( isNodeAbleToMakeProgress bootstrapPeersFlag
-                                       ledgerStateJudgement
-                                       hasOnlyBootstrapPeers
-            && targets' /= targets
-            && sanePeerSelectionTargets targets'
-            )
+      -- Possible values:
+      -- Nothing: churn governor is active
+      -- Just lsj: Previous ledger state judgement in Genesis mode, or
+      -- Just TooOld when not using Genesis
+      -- we can't use readTMVar because we will block when churn is
+      -- on thus making it useless.
+      mode <- tryReadTMVar churnMutex
+      targets' <- if useGenesisFlag && not (churning mode)
+                  then do
+                    swapTMVar churnMutex lsj
```

Comment:
not (churning mode) in the condition checks for this case, so churnMutex in fact will not be empty here.

Reply:
I see; we should add missing interfaces to `io-classes` & `strict-stm` - it's quite simple, and in the meantime before it's published to hackage one can always use a [`source-repository-package`](https://cabal.readthedocs.io/en/stable/cabal-project.html#specifying-packages-from-remote-version-control-locations).  We avoid to merge `s-r-p`s to `master`, but it's fine for development.

Reply:
All the info you need to make this decision is `useGenesisFlag` and `LSJ`, you have access to all that on the Churn side, and on the diffusion layer level, you can construct your STM actions to have this logic instead of the OG that should just read the right value and not care about this at all

Reply:
@bolt12 @coot 

I think there are two sides to this. The reason why I read the TVar in Genesis is that currentTargets is actually static information coming in from the configuration, but I still want to churn out peers. So maybe a better name for this is staticTargetsSelector. 
The second side is that you are right that maybe a mutex is not needed. So the churn governor does not have explicit argument with a LedgerStateJudgement, and so when I started work on this it looked to me that it doesn't know about the current ledger state. The risk I saw was interleaving of target values, but not in the sense you mentioned above of OG writing to the TVar, because I know it can't. What I had in mind is imagine this sequence in Genesis:
1. We are caught up, ie. TooYoung
1. Churn governor kicks in, and let's say manages to decrease active peers (its first action in go function), then delays for one second
2. Ledger state judgement changes to TooOld, targetPeers action fires and sets its private state target variable to the correct values from currentTargets (so Genesis targets)
3. Churn wakes up, and continues on, but because it uses the targets it discovered when the cycle began, it will perform its churn calculations with the wrong targets basis. It will write to the TVar with the decreased/increased targets from the TooYoung set.
4. OG will react to changes in TVar targets, and it will set its private target state with values from the TVar. Now we will end up with the wrong targets for Genesis (non-genesis targets).

Only later have I got to your commit 7509fdb in ouroboros-consensus, which ties LedgerStateJudgement with FetchMode, and so I can recover it in churn and not use the mutex. But now, every time I enter an atomic block when churning I will have to verify the current ledger state judgement, because it could have changed between the thread delays, and pull in the correct target basis from which to calculate new targets for the TVar. 

So I think a mutex is actually simpler here - at the start I take it in churn, let it continue setting it targets values to which the OG will react, and only at the end when there was a change in ledger state judgement, it will go through and OG will yet again active to set the correct targets.

I have much simplified this implementation, please stay tuned. 

Or perhaps I am still not seeing something?

Reply:
In step 3 the targets won't change for the churn governor, they will stay the same since you only change the internal state. It doesn't matter if you churn them if the governor doesn't notice that change that's another reason why the mutex doesn't solve anything, because in order for churn to be effective it needs to take into account `currentTargets` and the conditions to use those. 

You don't have to check everytime you enter an atomic block because there aren't any races going on. The only place where the targets Var is written is by the churn governor (I think).

Reply:
Is that not precisely the problem - because the OG does indeed notice the change because churn's static target basis stayed the same? Even in Genesis, the OG reads the TVar to set targets.
Yes, in step 3 the internal/private target in the state record does change in the OG which churn knows nothing about - that much is clear. But the problem is in the other direction. After step 4, churn will still use its non-Genesis/TooYoung targets as basis for calculations (because it only looks up the targets once, at beginning of every churn cycle), and it will send adjustments of those via TVar to the OG, which will notice: hey, new targets and they're different from the ones I have in my private set (the correct ones, which it just set a moment ago). This is regardless of whether or not Genesis is active. So it will take those targets from the TVar, and update its internal state to match them. So now the targets in the OG private state will reflect the targets from churn, and they will be wrong until the next cycle (or if ledger state changes yet again in the mean time). So at the end of the churning interval, the private target state in OG will be the values from TooYoung set in the example, where they should have been TooOld because we assumed that Genesis is active.

The race I see is with changes to the ledger state judgement that in turn fire off targetPeers (switching between static targets), which is also fired off when churning (adjustments to the static targets - but have to use the correct basis).

I don't mean to be argumentative and defend something that is wrong because I wrote it and wasting your time along the way - so pardon if this comes across like that. I genuinely see a problem here that a mutex can solve fairly cleanly- maybe I need a coffee or two.

Reply:
We can have a call and talk about his with more bandwidth but the issue is that you have the logic to decide on different targets on the OG and that depends on some external TVars. You want the Churn to be aware of that and expose a mutex from the OG to the CG (Churn Governor) so that the Churn Governor can do similar logic and churn on the correct targets. The mutex helps you dealing with what you want to do but I don't think it is the best design, because you are complicating things making the churn dependent on the OG which doesn't sound right. It is much simpler if you strive to keep them independent.

My suggestion is to put the deciding logic for the static targets on a different level of abstraction, for the OG you could put it at the `readPeerSelectionTargets :: STM m PeerSelectionTargets` level but the CG would still need to be aware of it. So you also have to account for it on CG. This works because the logic is static and depends on values that you can access above the stack. In the end you still have to have the Churn be aware of those variables (`useGenesis` and `lsj`) but for the OG it will just read the targets and not care about that logic.

Reply:
I would like to avoid a mutex; let's see ..., :thinking: but I also don't like that we have internal state and a `TVar` - ideally we need:

1. the internal state;
2. a way for the churn to read & modify it;
3. some form of an exclusive lock (not necessarily a mutex though, just a flag in the state who owns it (and thus can write to it).

The `TVar` solves point 2. - but it also makes its state available from two sources which is wired; if `churn` was part of OG then we wouldn't have that problem.

@bolt12 writing to a `TVar` which is not observed is also a code smell.

@bolt12 a design in which OG misses churn, is not good.  This is because churns are infrequent, so if it happens we want it to be effective. We have to make sure OG does not ignore it, churn should just postpone itself until it can run.

I'd prefer a solution which will be simpler to get us to the above ideal. Since we already have the `TVar` which the OG can ignore let see if we can design something that fits into our not ideal situation.

Both the OG and churn could compete for modifying the internal state, but we want only one of them to succeed. If we introduce a flag which I mentioned above, we can resolve the race. Since `churn` is outside of governor this will require one extra `TVar` which is monitored by the OG & can be read / written by the churn (if `churn` would be part of `OG` this would be simpler).

I also realise that such a flag right now will also need to be handled the way we do with targets, this needs to be stored in a `TVar`, because churn needs to be able to read & write & block on it, so this won't be that different from a `TMVar m ()` - so I think I can swallow it for now, unless we do it properly and bring churn into OG first and implement the changes on top of it; or one has even better idea?

Reply:
Yes, it is not ideal but OG but the OG not noticing the targets changing doesn't affect the overall design of both the OG and CG. If when in genesis mode we want to change the targets to some preset and that change be visible to every component that shares the targets TVar, then I defend that that should be done outside the OG. If this is respected CG and OG will be independent and not noticing the CG changes (in the case of bootstrap peerS) will be harmless.

The only place to touch the targets is the Churn Governor so if we now need something else to touch the targets maybe the best way is to have a new component `GenesisTargetGovernor` and have that component racing with the Churn governor instead of increasing the complexity between the CG and OG.

Reply:
That's an option, but not well aligned with including CG into OG.

Reply:
It will require access `LedgerStateJudgement`, and will need to be run in an extra thread, while this information is already maintained by OG - so it's a duplication of work.